qbit-manage: init at 4.6.5

flyingpeakock opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Added package qbit_manage. https://github.com/StuffAnThings/qbit_manage

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
qbit-manage: init at 4.6.5

flyingpeakock commented:
> Great effort! Can you give a bit more context about this:
> 
> > all qbit_manage config options must be defined or it will try to write to the config which will not work if it in the nix store.
> 
> All the available options seem a bit overwhelming at the moment. I'd prefer a set of necessary options and leave the rest for the user to input. But from your description it sounds like this is not an option?

If there are any missing options qbit_manage will try to add them with the default values which results in a crash. I have included setting the entire config to null in which case the user would configure qbit_manage outside of nix, and in this case the config would be writable.

By must be defined I don't mean by the user in their nix configuration, I mean in the module. Values with null in the module can stay as null and they will be written to the file and qbit_manage won't complain.

flyingpeakock commented:
Changed to copy config instead of symlinking it. Also removed most of the config options.

undefined-landmark reviewed (commented):
Left some comments. Overall very comprehensive (first?) contribution, nice. One part that is still missing is a test for the service module.

flyingpeakock commented on nixos/modules/services/torrent/qbit_manage.nix:
```diff
@@ -0,0 +1,583 @@
+
+          EnvironmentFile = mkIf (cfg.environmentFile != null) cfg.environmentFile;
+
+          # Hardening
```
Seems like setting DynamicUser implicitly sets some hardening options however I don't think its required here since I've manually set the hardening options.

flyingpeakock commented on pkgs/by-name/qb/qbit_manage/package.nix:
I've removed this line and it builds fine

undefined-landmark commented:
True, with the qbittorrent service it might be possible. I wrote the, maybe over extensive, tests for the qbittorrent module. Take a look there: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/qbittorrent.nix. Should give enough inspiration for how to setup a qbit instance :).

undefined-landmark reviewed (commented):
Some small comments. Currently already builds fine locally.

flyingpeakock commented on pkgs/by-name/qb/qbit_manage/package.nix:
done

flyingpeakock commented on pkgs/by-name/qb/qbit_manage/package.nix:
I had to add `dontCheckRuntimeDeps = true` to get it to build since upstream pins dependency versions. I tried `pythonRelaxDeps` first but it wasn't working for ruamel-yaml and bencodepy is called bencode-py in nix so that was also failing.

undefined-landmark commented:
Looks good! Only point remaining is this one imo: https://github.com/NixOS/nixpkgs/pull/466747#discussion_r2661682290. I'd prefer not using `dontCheckRuntimeDeps = true;` as the error it creates is correct. By turning it off completely we might miss other errors now or in the future. Some alternatives are mentioned in the referenced comment.

flyingpeakock commented:
I have submitted a PR to qbit_manage to change from bencodepy to bencode.py as well as included a patch which changes to bencode.py in case my PR doesn't get merged. The methods are the same for both of them so it is a drop in replacement.

undefined-landmark commented on pkgs/by-name/qb/qbit_manage/bencodepy.patch:
Nit: this patch can also be fetched with `fetchpatch`, so we don't need to store the patch in tree. But since this is such a small temporary patch it doesn't really matter all that much :).

undefined-landmark reviewed (approved):
Nice work! I think it's ready to merge now 👍. Thanks for your effort and all the adjustments.

flyingpeakock commented:
Thank you for the reviews, it has been very helpful.

SuperSandro2000 commented on pkgs/by-name/qb/qbit_manage/package.nix:
```diff
@@ -0,0 +1,75 @@
+  qbit_manage,
+}:
+python3Packages.buildPythonApplication rec {
+  pname = "qbit_manage";
```
```suggestion
  pname = "qbit-manage";
```

flyingpeakock replied:
done. Also changed the name of the parent folder.

SuperSandro2000 commented on pkgs/by-name/qb/qbit_manage/package.nix:
```diff
@@ -0,0 +1,75 @@
+              || (echo "Version mismatch: $outver" >&2; exit 1)
+            touch $out
+          '';
+    };
```
We have an inline hook for that.

flyingpeakock replied:
Done. I originally thought this wouldn't work since I have to set $HOME but I found a workaround.

SuperSandro2000 reviewed (commented):
The package should probably be named qbit-manage, too.

flyingpeakock commented:
Thanks for taking the time to review. I have now changed the package name and implemented your suggestions.
Sorry for taking so long, I was away on vacation.

flyingpeakock commented:
Changed the package name in the commit message

bbigras commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 466747`
Commit: `bab3de78fb47628247811c0b01301069dd3265e8`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>qbit-manage</li>
    <li>qbit-manage.dist</li>
  </ul>
</details>
